File: largest_rectangle.py
# // Time Complexity :O(n)
# // Space Complexity :O(n),stack space worst case
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :no
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        st=[]
        st.append(-1)
        maxar=0
        i=0
        while i <len(heights):
            
            if st[-1]==-1 or heights[i]>heights[st[-1]]:
                st.append(i)
            else:
                while st[-1]!=-1 and heights[i]<heights[st[-1]]:
                    top=st.pop()
                    arr=heights[top]*(i-st[-1]-1)
                    print(arr)
                    maxar=max(maxar,arr)
                st.append(i)
            i+=1
        while st[-1] != -1:
            top=st.pop()
            arr=heights[top]*(i-st[-1]-1)
            print(arr)
            maxar=max(maxar,arr)
            
        return maxar
                
        

File: Largest_Rectangle.java
// Time Complexity : O(n)
// Space Complexity : O(n) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int largestRectangleArea(int[] heights) {
        if(heights == null || heights.length == 0){
            return 0;
        }
        Stack<Integer> s = new Stack<>();
        int max = 0;
        int i = 0;
        s.push(-1);
        while(i < heights.length){
            if(s.peek() == -1 || heights[i] >= heights[s.peek()]){
                s.push(i);
                i++;
            }
            else{
                int popped = s.pop();
                max = Math.max(max, heights[popped] * (i - s.peek() -1)) ;
            }
        }
        while(s.size() > 1){
            int popped = s.pop();
            max = Math.max(max, heights[popped] * (i - s.peek() -1));
        }
        return max;
    }
}